libursa/src/encryption/symm/aescbc.rs

```diff
@@ -0,0 +1,141 @@
+
+        impl NewAead for $name {
+            type KeySize = $keysize;
+
+            fn new(key: GenericArray<u8, $keysize>) -> Self {
+                Self { key }
+            }
+        }
+
+        impl Aead for $name {
+            type NonceSize = $noncesize;
+            type TagSize = $tagsize;
+            type CiphertextOverhead = U0;
+
+            fn encrypt<'msg, 'aad>(
+                &self,
+                nonce: &GenericArray<u8, Self::NonceSize>,
+                plaintext: impl Into<Payload<'msg, 'aad>>,
+            ) -> Result<Vec<u8>, Error> {
+                let payload = plaintext.into();
+                let encryptor =
+                    $algo::new_var(&self.key[..$algokeysize::to_usize()], &nonce.as_slice())
+                        .map_err(|_| Error)?;
+                let mut ciphertext = encryptor.encrypt_vec(payload.msg);
+                let mut hmac =
+                    $hash::new_varkey(&self.key[$algokeysize::to_usize()..]).map_err(|_| Error)?;
+                hmac.input(payload.aad);
+                hmac.input(nonce.as_slice());
+                hmac.input(ciphertext.as_slice());
+                let hash = hmac.result().code();
+                ciphertext.extend_from_slice(hash.as_slice());
+                Ok(ciphertext)
+            }
+
+            fn decrypt<'msg, 'aad>(
+                &self,
+                nonce: &GenericArray<u8, Self::NonceSize>,
+                ciphertext: impl Into<Payload<'msg, 'aad>>,
+            ) -> Result<Vec<u8>, Error> {
+                let payload = ciphertext.into();
+
+                if payload.msg.len() < Self::TagSize::to_usize() + $algokeysize::to_usize() {
+                    return Err(Error);
+                }
+
+                let tag_start = payload.msg.len() - Self::TagSize::to_usize();
+                let buffer = Vec::from(&payload.msg[..tag_start]);
+                let tag = Vec::from(&payload.msg[tag_start..]);
+
+                let mut hmac =
+                    $hash::new_varkey(&self.key[$algokeysize::to_usize()..]).map_err(|_| Error)?;
+                hmac.input(payload.aad);
+                hmac.input(nonce.as_slice());
+                hmac.input(buffer.as_slice());
+                let expected_tag = hmac.result().code();
+
+                if expected_tag.ct_eq(&tag).unwrap_u8() == 1 {
+                    let decryptor =
+                        $algo::new_var(&self.key[..$algokeysize::to_usize()], &nonce.as_slice())
+                            .map_err(|_| Error)?;
+                    let plaintext = decryptor
+                        .decrypt_vec(buffer.as_slice())
+                        .map_err(|_| Error)?;
+                    Ok(plaintext)
+                } else {
+                    Err(Error)
+                }
+            }
+        }
+
+        default_impl!($name);
+        drop_impl!($name);
+        serialize_impl!($name, $visitor);
+    };
+}
+
+aes_cbc_hmac_impl!(
+    Aes128CbcHmac256,
+    U16,
```

Comment:
I see these `U` values defined in http://fizyk20.github.io/generic-array/typenum/consts/index.html but what is the objective here?